examples/command.roc

```diff
@@ -1,13 +1,19 @@
-app [main] { pf: platform "../platform/main.roc" }
+app [Model, server] { pf: platform "../platform/main.roc" }
 
 import pf.Task exposing [Task]
 import pf.Http exposing [Request, Response]
 import pf.Command
 import pf.Utc
 
-main : Request -> Task Response []
-main = \req ->
+Model : {}
 
+server = { init, respond }
+
+init : Task Model [Exit I32 Str]_
+init = Task.ok {}
```

Comment:
It's probably best if we leave this for another PR but I'd like to minimize the boiler plate here some time. Perhaps we could make server a tag union like `[NoInit (Http.Request -> Task Http.Response [ServerErr Str]_), WithInit { init: _, respond: _}]`

Reply:
I was looking at this change and it makes some of the state management and juggling quite a bit more complex in the platform. I would prefer to delay this to a later PR.

Also, this is minimal boilerplate currently.
```
Model : {}
server = { init: Task.ok {}, respond }

respond : Request, Model -> Task Response [ServerErr Str]_
respond = \req, _ ->
    # Log request datetime, method and url
    datetime = Utc.now! |> Utc.toIso8601Str

    Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

    Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" }
```

Reply:
Sorry to "interrupt"

What about making a `createServer` function that uses optional record params? It'd allow omitting the `init`, and potentially let you hide the implementation a bit (if you put it behind an opaque type, otherwise it's just like a JS-style `defineConfig`).
Slightly less verbose, about what it is now, but scales (just add an init).

Maybe I'm missing something, but it seems to me like that'd work.

Reply:
So part of the complication is that the app defines the model and the platform doesn't know what type it is.

So the init function is a `Task (Box Model) I32`. Having a default branch in the platform that generates a `Task.ok {}` for the model, will not unify if the model is not a `{}`. So the types are definitely a bit more finicky here.

I need to mess around with this more though. I'm sure there must be a clean way to make this work.

Reply:
Ah, yeah, that makes sense. I figured the user's model might mess things up. I don't think the type system would let you do this, but in my head you'd be able to default the model to `{}` and force it define init if it's not `{}`, as you'd need to define it if it's not nothing anyway, right?
I don't _think_ you can do it, but that's the API I imagine people would like. It's what I'd like, at least.

Reply:
Also inconvenient, `Model` has to be specified no matter what. Roc doesn't have a way for us to optionally require a type be passed from the app to the platform. So even if we have a config function, it would be at a minimum something like:

```roc
Model : {}

server = noInit \req ->
    # Log request datetime, method and url
    datetime = Utc.now! |> Utc.toIso8601Str

    Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

    Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" }
```

vs:
```roc
Model : {}

server = {
    init: Task.ok {},
    respond: \req, _ ->
        # Log request datetime, method and url
        datetime = Utc.now! |> Utc.toIso8601Str

        Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

        Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" },
}
```

So a tiny bit denser, but not a ton. Though, I definitely wouldn't nest respond in the server record directly. I don't think it is great for readabilty. So more realistically, it would be:
```roc
Model : {}

server = { init: Task.ok {}, respond }
respond = \req, _ ->
    # Log request datetime, method and url
    datetime = Utc.now! |> Utc.toIso8601Str

    Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"

    Task.ok { status: 200, headers: [], body: Str.toUtf8 "<b>Hello, world!</b></br>" }
```

So only one extra line of code.

Reply:
Yeah. At that point, it's not worth it, _unless_ you want to be able to add subscriptions/whateverelseElmhas in a non-breaking manner, but at that point, you're probably breaking other things anyway.

Reply:
> I don't think you can do it, but that's the API I imagine people would like. It's what I'd like, at least.

Yeah, there is probably a way to make the api work and be nicer, but I haven't quite found it yet. At least not without adding a bunch of extra conditionals to the platform.

Like we can make the `server` be a tag union `[ WithInit { init, respond }, NoInit respond ]`, but the users still has to specify a `Model` due to how roc works. Also, the platform has to do a lot of extra conditionals to make the types happy. It can no longer use `Model` directly. It has to use something like `[Model Model, NoModel]`. Just adds conditionals that feel very unnecessary.

Reply:
Yeah, and the tag union looks ugly. I'd just leave it be, it can be improved some other time.

Though, why can't you extract model out? i.e., rather than passing it, just make the defineConfig-y function take a typeparam, then extract the typeparam. Or does the typesystem not allow that?

Reply:
I feel like I did something similar in the past, but I wasn't having luck when I attempted to do so today. Maybe I am just formulating something wrong.